### PR TITLE
Add tags support to signage

### PR DESCRIPTION
**Pull request by vache**
Signs defined in json mapgen will now support <full_name>, <given_name>, and <family_name> tags that can be used to output a random name, as well as a <city> tag that can be used to output the nearest city name.

Right now, signs are the only feature that supports these tags, but the intention is to eventually extend them to be able to be used in NPC conversations and elsewhere.  

Also fixes a small issue with the <npcname> replacement that allows using multiple <npcname> tags in the same line to refer to the same NPC.

### CODE REVIEW

File: src/overmap.cpp [hunk 1]
```diff
@@ -4369,6 +4369,21 @@ void overmap::add_mon_group(const mongroup &group)
     DebugLog( D_ERROR, D_GAME ) << group.type.str() << ": " << group.population << " => " << xpop;
 }
 
+const city& overmap::nearest_city(int x, int y)
```
Comment (BevapDin) [inline src/overmap.cpp]: Why not `overmap_buffer.closest_city`? It works across overmap borders.

File: src/output.cpp [hunk 2]
```diff
@@ -1720,6 +1721,63 @@ std::string string_format(std::string pattern, ...)
...
+    input = replace_given_name_tag(input);
+    return input;
+}
+
+std::string replace_city_tag(std::string input, std::string name)
+{
+    size_t offset = input.find("<city>");
+    while (offset != std::string::npos) {
+        input.replace(offset, 6, name);
+        // capitalize the first letter of the city
```
Comment (BevapDin) [inline src/output.cpp]: Why all the capitalization? Names should already have the proper form, there should be no need to do this manually. Btw. the original code in `replace_with_npc_name` capitalized the first character of the *sentence*.

Also: holy code duplication! Have you considered to create a simple `void replace(std::string &text, const std::string &pattern, const std::string &replacement); ` function that replaces all appearances of string A with string B in a given input. And just call that function `replace(text, "<city>", city_name)`?

That function could be used in many other places.

File: src/mapgen.cpp [hunk 3]
```diff
@@ -663,7 +663,35 @@ class jmapgen_sign : public jmapgen_piece {
...
+        std::string signtext;
+
+        if( !snippet.empty() ) {
+            // select a snippet from the category
+            signtext = SNIPPET.get( SNIPPET.assign( snippet ) );
+        } else if( !signage.empty() ) {
+            signtext = signage;
+        } else {
+            // throw some sort of error
+            debugmsg("jmapgen_sign with no signage or snippet defined");
```
Comment (BevapDin) [inline src/mapgen.cpp]: That should probably be in the constructor. The failure is easy to detect, and it should be shown as soon as possible. You can even use the json error reporting from there: `jsi.throw_error("need either X or Y, but found neither")`

**Comment by chaosvolt** [comment]
Neat.

**Comment by vache** [comment]
Integrated all of BevapDin's comments.  Did not address any other situations where the replace_substring might be useful though.

**Comment by Zireael07** [comment]
Can we have city signs somewhere?

**Comment by vache** [comment]
That's something I want to tackle eventually.  The idea is to have a chance to put road signs just outside cities saying something like "Aberdeen City Limits" or "Entering Avondale, Pop. 2135" etc.  It is probably going to take some tweaking of the city generation code to get something like that placed right though.

**Pull request merged**